dipy/align/vector_fields.pyx

```diff
@@ -3063,9 +3066,65 @@ def create_sphere(cnp.npy_intp nslices, cnp.npy_intp nrows,
     return np.asarray(s)
 
 
+cdef void _gradient_3d_fun(floating[:, :, :] img, double[:, :] img_world2grid,
+                 double[:] img_spacing, double[:, :] out_grid2world,
+                 floating[:, :, :, :] out, int[:, :, :] inside, int k) nogil:
+    cdef:
+        int nrows = out.shape[1]
+        int ncols = out.shape[2]
+        int i, j, in_flag, p
+        double tmp, x[3], dx[3], h[3], q[3]
+    h[0] = 0.5 * img_spacing[0]
+    h[1] = 0.5 * img_spacing[1]
+    h[2] = 0.5 * img_spacing[2]
+    for i in range(nrows):
+        for j in range(ncols):
+            inside[k, i, j] = 1
+            # Compute coordinates of index (k, i, j) in physical space
+            x[0] = _apply_affine_3d_x0(k, i, j, 1, out_grid2world)
+            x[1] = _apply_affine_3d_x1(k, i, j, 1, out_grid2world)
+            x[2] = _apply_affine_3d_x2(k, i, j, 1, out_grid2world)
+            for p in range(3):
+                dx[p] = x[p]
+            for p in range(3):
+                # Compute coordinates of point dx on img's grid
+                dx[p] = x[p] - h[p]
+                q[0] = _apply_affine_3d_x0(dx[0], dx[1], dx[2], 1,
```

Comment:
would it be possible to do an internal loop with mapping to the `_apply_affine_3d_x[0,1,2]`